anoma/namada-interface
# feat./1081 - Namadillo: Check Tx applied status when broadcasting

Resolves #1081 

Using the response from `rpc.broadcastTx()`, check the `isApplied` value for each commitment, raise error if any are not applied.

- [x] Update SDK to fix issue where we were not getting accurate `isApplied` values if any of the innerTx were not applied
- [x] Export SDK function to return the computed inner Tx hashes of all commitments (this is used later for mapping to inner Tx props)
- [x] Update Toast & Notifications in Namadillo to collect successful/failed Tx and display a new notification to users, `PartialSuccess`, where both successful and failed inner-Tx can be displayed

This makes use of the `ProcessTxResponse` from the SDK's `rpc.broadcastTx()`. Example:
```json
{
    "code": "0",
    "commitments": [
        {
            "hash": "DE979B02CC9014068E09DEFCB26592FB2A992D43B35B57C1CC9643BAE4F392CE",
            "isApplied": true
        },
        {
            "hash": "C91EED4782248D482517AE5C37AEA9895E1469439C42C059789C334FB9F8031C",
            "isApplied": true
        }
    ],
    "gasUsed": "210",
    "hash": "BC99E787272BDD15468E3F5C86EEC2FBAE752C568983086805EDB5286F5BDB95",
    "height": "16602",
    "info": "Check batch for result.",
    "log": ""
}
```

Each commitment in this response has a boolean indicating whether that commitment has been accepted by all VPs and applied on chain.

### Testing

Testing this PR is a bit weird, as you need to trigger a VP rejection _after_ SDK validation passes. An easy way to do this is to hijack the `gasLimit` value. In Namadillo, in `src/lib/query.ts` -> `getTxProps`:
- use (for example)`gasLimit: BigNumber(100)`, then submit 3 staking Tx - all should fail
- Set this value to `BigNumber(285)` and only one will fail.

**NOTE** You can test the results (check the hashes) in the error msg against what is in the console

### Screenshot

Partial success notification, indicates both successful and unsuccessful Tx:
<img width="400" alt="Screen Shot 2024-09-10 at 3 36 34 PM" src="https://github.com/user-attachments/assets/c8330037-a1fc-4508-aa5e-8af25495d3de">

Failed state when SDK validation passes, but the Tx response indicates that commitments were rejected by VPs:
<img width="435" alt="Screen Shot 2024-09-10 at 4 57 21 PM" src="https://github.com/user-attachments/assets/6dcdf4c7-fce6-44a9-8725-34d7b0dfeb9f">


Success notification is unchanged.
